tools/ReadConsoleInputStream/ConcurrentCircularQueue.cs

```diff
@@ -0,0 +1,62 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+
+namespace Nivot.Terminal
+{
+    /// <summary>
+    /// Implements a circular buffer.
+    /// </summary>
+    /// <typeparam name="T"></typeparam>
+    public class ConcurrentCircularQueue<T> : ConcurrentQueue<T>
+    {
+        public ConcurrentCircularQueue(int capacity)
+        {
+            Capacity = GetAlignedCapacity(capacity);
+        }
+
+        /// <summary>
+        /// 
+        /// </summary>
+        /// <param name="collection"></param>
+        /// <param name="capacity"></param>
+        public ConcurrentCircularQueue(IEnumerable<T> collection, int capacity) : base(collection)
+        {
+            Capacity = GetAlignedCapacity(capacity);
+        }
+
+        private int GetAlignedCapacity(int n)
+        {
+            if (n < 2)
+            {
+                throw new ArgumentException("Capacity must be at least 2");
+            }
+
+            var f = Math.Log(n, 2);
+            var p = Math.Ceiling(f);
+
+            return (int) Math.Pow(2, p);
+        }
+
+        public new void Enqueue(T item)
+        {
+            if (Count >= Capacity)
+            {
+                lock (this)
+                {
+                    while (Count >= Capacity)
+                    {
+                        TryDequeue(out _);
+                    }
```

Comment:
Not familiar with this syntax. Does this mean that if the queue is too full, you're going to toss the oldest items out into nowhere? Maybe leave a comment as such.

Reply:
Yep, sorry, it's funny -- looking at it now, I had originally intended to create a circular queue, but I ended up writing a bounded queue instead but I had not stopped to reconsider the naming. The three hardest things in coding: naming things and off-by-one errors -- right? :)

Reply:
Absolutely. :P